---
title: "Constructing Metadata Files"
author: "Alicia Rich"
output:
  html_document:
    toc: true
    toc_location: "before"
    toc_depth: 4
    number_sections: false
    toc_float: true
    code_folding: "hide"
    fig_caption: true
editor_options: 
  chunk_output_type: inline
params:
  sampleset: "loris"
                     
---

```{r, include = F}
global             <- config::get(config = "default")

source(paste0(global$setup))
source(paste0(swan$functions))

seqruns <- list(
  loris     = as.list(paste0("hdz", 1:18)),
  marmoset  = as.list(sprintf("cm%03d", 1:10))
)

subject_list <- keep_at(subjects, paste0(params$sampleset)) %>% list_flatten(name_spec = "{inner}")
path               <- config::get(config = params$sampleset)

source(path$metadata$bristol)
source(path$metadata$key)

opts_chunk$set(message = FALSE, warning = FALSE)
```


# Intro

The purpose of a metadata file is to organize the potential independent or predictor variables for your analysis into a single table with one row per SampleID. Then, when you produce a set of potential dependent or outcome values, you organize those into a similar structure with the SampleIDs organized rowwise to streamline the process of matching predictor variables to outcome variables by SampleID. It's good practice to keep as much of your information in one tidy table as possible so that you can keep pulling from that source to further filter, wrangle and analyze without losing track of different versions and datasets over time. That means you should brainstorm as many possible predictor variables you might use in downstream analysis as possible and organize them into one tidy table where each SampleID is matched to a value for every variable. You will end up ignoring most of these variables as you construct individual tests and visuals later, so consider this simply a rough draft of your information of interest. I am going to do this with the Pygmy Loris dataset for this tutorial. You may have a very different set of variables to organize for your own project.  

## Previous Scripts

Before this, you should use the SampleInventory script to create the file compilation.tsv.  

## Configurations

### Sampleset in Params

You can use the sampleset setting under params in the header of this script to select which sampleset you will be working with. So long as the same name is used consistently, this should automatically filter for that name (e.g., loris or marmoset). 

### Sequencing Run Lists

Below is a chunk to generate a list of formatted codes for each available sequencing run to date, separated by taxa/samplesets (currently just for loris and marmoset). Make sure the end number matches the highest integer we have for that sampleset to date. The next chunk will format a list of all subject names for whichever taxon you identified in the header.

```{r}
seqruns <- list(
  loris     = as.list(paste0("hdz", 1:18)),
  marmoset  = as.list(sprintf("cm%03d", 1:10))
)

subject_list <- keep_at(subjects, paste0(params$sampleset)) %>% list_flatten(name_spec = "{inner}")
```

### File Paths

Next, you should make sure your config.yml file contains the path to locate each of the files you will be using. Below is an example excerpt from my config file. 

```
sample_sheets:
  compilations:
    loris:    "../bioinformatics_stats/dataframes/sample_sheet/loris/hdz_combined_sample_sheet.csv"
    marmoset: "../bioinformatics_stats/dataframes/sample_sheet/marmoset/cm_combined_sample_sheet.csv"
  hdz1:  "../bioinformatics_stats/dataframes/sample_sheet/loris/hdz1_sample_sheet.csv"
  hdz2:  "../bioinformatics_stats/dataframes/sample_sheet/loris/hdz2_sample_sheet.csv"
  hdz3:  "../bioinformatics_stats/dataframes/sample_sheet/loris/hdz3_sample_sheet.csv"
  cm001: "../bioinformatics_stats/dataframes/sample_sheet/marmoset/cm001_sample_sheet.csv"
  cm002: "../bioinformatics_stats/dataframes/sample_sheet/marmoset/cm002_sample_sheet.csv"
  cm003: "../bioinformatics_stats/dataframes/sample_sheet/marmoset/cm003_sample_sheet.csv"

barcode_alignments:
  compilations:
    loris: "../labwork/minion_data/barcode_alignments/loris/hdz_combined_barcode_alignment.tsv"
    marmoset: "../labwork/minion_data/barcode_alignments/marmoset/cm_combined_barcode_alignment.tsv"
  hdz1:  "../labwork/minion_data/barcode_alignments/loris/hdz1_barcode_alignment.tsv"
  hdz2:  "../labwork/minion_data/barcode_alignments/loris/hdz2_barcode_alignment.tsv"
  hdz3:  "../labwork/minion_data/barcode_alignments/loris/hdz3_barcode_alignment.tsv"
  cm001: "../labwork/minion_data/barcode_alignments/marmoset/cm001_barcode_alignment.tsv"
  cm002: "../labwork/minion_data/barcode_alignments/marmoset/cm002_barcode_alignment.tsv"
  cm003: "../labwork/minion_data/barcode_alignments/marmoset/cm003_barcode_alignment.tsv"
  
loris:
  libraries_csv: "../data/libraries_loris.csv"
  compilation_csv: "../data/compilation_loris.csv"
  extracts_csv: "../data/extracts_loris.csv"
  samples_csv: "../data/samples_loris.csv"
  sample_sheet: "../bioinformatics_stats/loris/dataframes/sample_sheet/"
  taxonomy_list: "../bioinformatics_stats/loris/dataframes/taxonomy_list/"

marmoset:
  libraries_csv: "../data/libraries_marmoset.csv"
  compilation_csv: "../data/compilation_marmoset.csv"
  extracts_csv: "../data/extracts_marmoset.csv"
  samples_csv: "../data/samples_marmoset.csv"
  sample_sheet: "../bioinformatics_stats/marmoset/dataframes/sample_sheet/"
  taxonomy_list: "../bioinformatics_stats/marmoset/dataframes/taxonomy_list/"

```

Note that I also include paths to files that this script will create. If the file is already there, then it will be overwritten, if not, it will be created there. Run the code below to set up your paths from the config file for the working sampleset you identified in the header:

```{r, message = FALSE}
path               <- config::get(config = params$sampleset)
```

### Other Scripts

I also create some of the tables and lists used here in some scripts that the chunk below will run. This just keeps this script a bit tidier by modularizing some of the parts.

```{r, message = FALSE}
source(path$metadata$bristol)
```


# Script

The chunk below is duplicated above. I am adding it here to run the full setup chunk again in case you skipped past the intro material:

```{r, message = FALSE}
global             <- config::get(config = "default")

source(paste0(global$setup))
source(paste0(swan$functions))

seqruns <- list(
  loris     = as.list(paste0("hdz", 1:18)),
  marmoset  = as.list(sprintf("cm%03d", 1:10))
)

subject_list <- keep_at(subjects, paste0(params$sampleset)) %>% list_flatten(name_spec = "{inner}")
path               <- config::get(config = params$sampleset)

source(path$metadata$bristol)
```

## Sample Inventory

First I will read in the sample inventory, which is actually already a metadata table itself. We just want to match some more interesting variables to it.

```{r}
samples <- read.table(path$compilation, header = T, sep = "\t") %>%
            mutate(CollectionDate = ymd(CollectionDate))
```

## Organizing Dates

The toughest variables to match and wrange in R are often dates, especially when you are dealing with both states (start and end dates for intervals) and events (single dates). For studies like this one based around daily sample collection, I find it easiest to start by populating a blank dataframe with one row per sample collection day. Then I match date-based variables to this dataframe and join my SampleIDs to it.

```{r}
study_days <- tibble(Date = seq.Date(from = ymd(path$day1), to = ymd(path$last), by = "day")) %>%
  mutate(study_day = row_number())
```

## Construct Variables

### Diet

We have only two subjects in this dataset: Culi and Warble. Culi underwent a series of diet trials that we organized into start and end dates, while Warble maintained a baseline diet throughout our study.

```{r}
# steroid doses are calculated as mg/day while probiotic and oatgel values are binary (included = 1, none = 0)
diet_trials   <- list(
baseline      = tibble(steroid = 0.00, probiotic = 0, oatgel = 0),
option1       = tibble(steroid = 0.00, probiotic = 1, oatgel = 0),
option2       = tibble(steroid = 0.00, probiotic = 1, oatgel = 1),
option3       = tibble(steroid = 0.10, probiotic = 0, oatgel = 0),
option4       = tibble(steroid = 0.20, probiotic = 0, oatgel = 0),
option5       = tibble(steroid = 0.20, probiotic = 1, oatgel = 1),
option6       = tibble(steroid = 0.20, probiotic = 0, oatgel = 1),
option7       = tibble(steroid = 0.10, probiotic = 1, oatgel = 1),
option8       = tibble(steroid = 0.05, probiotic = 1, oatgel = 1)
)
```

```{r, message = FALSE}
# optionals are foods that are routinely cycled in and out of daily diets to maximize variety. here I am only noting those optionals which were permanently removed from culi's diet at a certain date in an attempt to reduce GI inflammation.
optionals.culi <- list(
  tomatoes    = tibble(start_day = ymd(path$day1), last_day = ymd("2024-02-19")),
  cauliflower = tibble(start_day = ymd(path$day1), last_day = ymd("2024-01-22")),
  broccoli    = tibble(start_day = ymd(path$day1), last_day = ymd("2024-01-22"))
) 
  
optionals.culi <-   imodify(optionals.culi, ~ {
  food <- .y
  tbl  <- .x
  rowwise(tbl) %>%
    mutate(Date = list(seq.Date(from = start_day, to = last_day, by = "day"))) %>%
  unnest(Date) %>% mutate(food = food) %>% select(Date, food) 
}) %>%
  list_rbind() %>%
  mutate(value = 1) %>%
  pivot_wider(
    id_cols     = Date,
    names_from  = food,
    values_fill = 0
  ) %>%
  right_join(select(study_days, Date)) %>%
  mutate(across(!Date, ~ replace_na(.x, 0))) %>%
  nest(.by = Date, .key = "Foods_included")
```

```{r}
culi.trials <- tribble(
~start_day,        ~last_day        , ~diet,
ymd(loris$day1)  , ymd("2023-11-01"), diet_trials$baseline,
ymd("2023-11-02"), ymd("2023-11-22"), diet_trials$option1,
ymd("2023-11-23"), ymd("2023-12-13"), diet_trials$option2,
ymd("2023-12-14"), ymd("2024-01-03"), diet_trials$option3,
ymd("2024-01-04"), ymd("2024-01-24"), diet_trials$option4,
ymd("2024-01-25"), ymd("2024-02-19"), diet_trials$option5,
ymd("2024-02-20"), ymd("2024-02-22"), diet_trials$option6,
ymd("2024-02-23"), ymd("2024-05-14"), diet_trials$option7,
ymd("2024-05-15"), ymd(path$last)   , diet_trials$option8
)

culi.schedule <- culi.trials %>%
  rowwise() %>%
  mutate(Date = list(seq.Date(from = start_day, to = last_day, by = "day"))) %>%
  unnest(Date) %>%
  select(Date, 
         Trial_Phase = diet) %>%
  left_join(optionals.culi, by = join_by(Date)) %>%
  unnest(c(Trial_Phase, Foods_included))
```

```{r}
culi.diet.schedule <- list(
baseline = tibble(start_day = ymd(loris$day1)  , last_day = ymd("2023-11-01"), diet = diet_trials$baseline),
phase1   = tibble(start_day = ymd("2023-11-02"), last_day = ymd("2023-11-22"), diet = diet_trials$option1),
phase2   = tibble(start_day = ymd("2023-11-23"), last_day = ymd("2023-12-13"), diet = diet_trials$option2),
phase3   = tibble(start_day = ymd("2023-12-14"), last_day = ymd("2024-01-03"), diet = diet_trials$option3),
phase4   = tibble(start_day = ymd("2024-01-04"), last_day = ymd("2024-01-24"), diet = diet_trials$option4),
phase5   = tibble(start_day = ymd("2024-01-25"), last_day = ymd("2024-02-19"), diet = diet_trials$option5),
phase6   = tibble(start_day = ymd("2024-02-20"), last_day = ymd("2024-02-22"), diet = diet_trials$option6),
phase7   = tibble(start_day = ymd("2024-02-23"), last_day = ymd("2024-05-14"), diet = diet_trials$option7),
phase8   = tibble(start_day = ymd("2024-05-15"), last_day = ymd(path$last)   , diet = diet_trials$option8)
)
```

```{r}
culi.diet.schedule %>%
  list_rbind() %>%
  mutate(Probiotic   = if_else(diet$probiotic == 1, "check", "minus"),
         Oatgel      = if_else(diet$oatgel == 1, "check", "minus"),
         Steroid     = if_else(diet$steroid == 0.00, "minus", as.character(diet$steroid))) %>%
  select(start_day,
         Steroid,
         Probiotic,
         Oatgel) %>%
  gt() %>%
    fmt_date(columns = ends_with("_day"), date_style = "m_day_year") %>%
    cols_label(start_day = "Begin") %>%
    fmt_icon( columns = Steroid, rows = Steroid == "minus") %>%
    fmt_icon( columns = c("Probiotic", "Oatgel")) %>%
    cols_align("center") %>%
    tab_header("Summary of Trial Phases for Culi") %>%
    tab_footnote("Phase ends on the date the next row begins.", 
                 locations = cells_column_labels(start_day)) %>%
    tab_footnote("Steroid (Budesonide) dose reported as mg/day", 
                 locations = cells_column_labels(Steroid)) %>%
    tab_footnote("Warble stayed at baseline for duration of the study.", 
                 locations = cells_title()) %>%
    opt_stylize(style = 3, color = "gray")
```


```{r}
optionals.warb <- study_days %>%
  select(Date) %>%
  mutate(tomatoes = 1, cauliflower = 1, broccoli = 1)

warble.schedule <- study_days %>%
  select(Date) %>%
  rowwise() %>%
  cbind(diet_trials$baseline) %>%
  left_join(optionals.warb, by = join_by(Date))
  
```

### Medication

In addition to the steroids included in trials, Culi recieved antibiotics for a stint during our collections. We definitely want to keep something like antibiotics factored into our microbiome data analysis, so I will create a table for that.

```{r}
meds <- list(
  culi = tribble(
~start_day,        ~last_day        , ~med_type,  ~med_name,  ~med_dose, ~dose_units,
ymd(path$day1), ymd("2024-04-28"), "none", "none", 0, "mg/day",
ymd("2024-04-29"), ymd("2024-05-06"), "antibiotic", "Metronidazole", 12.5, "mg/day",
ymd("2024-05-06"), ymd(path$last), "none", "none", 0, "mg/day"
  ),
  warble = tribble(
~start_day,        ~last_day        , ~med_type,  ~med_name,  ~med_dose, ~dose_units,
ymd(path$day1), ymd(path$last), "none", "none", 0, "mg/day"
  )
)

culi.meds <- meds$culi %>%
  rowwise() %>%
  mutate(Date = list(seq.Date(from = start_day, to = last_day, by = "day"))) %>%
  unnest(Date) %>%
  select(Date, 
         med_type,
         med_name,
         med_dose,
         dose_units
         )

warble.meds <- meds$warble %>%
  rowwise() %>%
  mutate(Date = list(seq.Date(from = start_day, to = last_day, by = "day"))) %>%
  unnest(Date) %>%
  select(Date, 
         med_type,
         med_name,
         med_dose,
         dose_units
         )
  
```


### Housing

Over the course of the study, both lorises moved from neighboring enclosures in one building to a much more expansive set of enclosures with more variable shifting capabilities in another building. They moved at different dates, so I want to score each sample based on those different date ranges.

```{r}
housing <- list(
  warble = tribble(
~start_day,        ~last_day        , ~enclosure,
ymd(loris$day1)    , ymd("2024-02-28"), "old",
ymd("2024-02-29")  , ymd(loris$last)  , "new"
  ),
  culi = tribble(
~start_day,        ~last_day        , ~enclosure,
ymd(loris$day1)    , ymd("2024-02-27"), "old",
ymd("2024-02-28")  , ymd(loris$last)  , "new"
  )
)

warble.housing <- housing$warble %>%
  rowwise() %>%
  mutate(Date = list(seq.Date(from = start_day, to = last_day, by = "day"))) %>%
  unnest(Date) %>%
  select(Date, 
         enclosure)

culi.housing <- housing$culi %>%
  rowwise() %>%
  mutate(Date = list(seq.Date(from = start_day, to = last_day, by = "day"))) %>%
  unnest(Date) %>%
  select(Date, 
         enclosure)
```


### Bristol Fecal Scores

I only have scores available for Culi, and scores were recorded anywhere from 0 to 3x per day.

```{r}
culi.bristol <- bristol %>%
  group_by(date) %>%
  summarize(bristol_min  = min(score),
            bristol_mean = round(mean(score), digits = 0),
            bristol_max  = max(score)) %>%
  ungroup() %>%
  full_join(study_days, by = join_by(date == Date)) %>%
  select(Date = date, 
         bristol_min,
         bristol_mean,
         bristol_max) %>%
  arrange(Date)
```

### Repro

```{r}
repro.log    <- list(
  access = tribble(
    ~value,     ~start_day,  ~last_day,
    1     , ymd(path$day1)   , ymd("2023-10-29"),
    0     , ymd("2023-10-30"), ymd("2024-04-01"),
    1     , ymd("2024-04-02"), ymd("2024-04-15"),
    0     , ymd("2024-04-16"), ymd(path$last)
  ),
  estrus = tribble(
    ~value,     ~start_day,  ~last_day,
    1     , ymd(path$day1)   , ymd("2023-10-29"),
    0     , ymd("2023-10-30"), ymd("2024-03-21"),
    1     , ymd("2024-03-21"), ymd("2024-04-05"),
    0     , ymd("2024-04-06"), ymd(path$last)
  ),
  pregnant = tribble(
    ~value,     ~start_day,  ~last_day,
    0     , ymd(path$day1)   , ymd("2024-04-05"),
    1     , ymd("2024-04-06"), ymd("2024-10-08"),
    0     , ymd("2024-10-09"), ymd(path$last)
  ),
  notes = tribble(
    ~repro_note          , ~subject  ,  ~Date,
    "swollen_testicles"  , "culi"    , ymd("2024-02-29"),
    "swollen_vulva"      , "warble"  , ymd("2024-03-21"),
    "whistling"          , "culi"    , ymd("2024-03-21"),
    "whistling"          , "warble"  , ymd("2024-04-04"),
    "copulation_observed", "both"    , ymd("2024-04-03"),
    "copulation_observed", "both"    , ymd("2024-04-04"),
    "sperm_plug_observed", "warble"  , ymd("2024-04-05"),
    "birth"              , "warble"  , ymd("2024-10-09"),
    "double infanticide" , "warble"  , ymd("2024-10-10")
  )
) 
```

```{r, warning = FALSE}
access.schedule <- repro.log$access %>%
  rowwise() %>%
  mutate(Date = list(seq.Date(from = start_day, to = last_day, by = "day"))) %>%
  unnest(Date) %>%
  select(Date, 
         access = value)

estrus.schedule <- repro.log$estrus %>%
  rowwise() %>%
  mutate(Date = list(seq.Date(from = start_day, to = last_day, by = "day"))) %>%
  unnest(Date) %>%
  select(Date, 
         estrus = value)


preg.schedule <- repro.log$pregnant %>%
  rowwise() %>%
  mutate(Date = list(seq.Date(from = start_day, to = last_day, by = "day"))) %>%
  unnest(Date) %>%
  select(Date, 
         pregnant = value)

warble.repro.notes <- repro.log$notes %>%
  filter(subject == "warble" | subject == "both") %>%
  select(Date, 
         repro_note) %>%
  group_by(Date) %>%
  summarise(repro_note = str_c(repro_note, collapse = "; "), .groups = "drop")

culi.repro.notes <- repro.log$notes %>%
  filter(subject == "culi" | subject == "both") %>%
  select(Date, 
         repro_note) %>%
  group_by(Date) %>%
  summarise(repro_note = str_c(repro_note, collapse = "; "), .groups = "drop")
  
```


### Health

Now I will score the misc. notes provided by staff on health observations as variables. 

```{r}
culi.health <- tribble(
  ~health_note                   , ~Date,
  "dried blood around anus/tail" , ymd("2023-11-16"),
  "coprophagy observed"          , ymd("2023-12-17"),
  "weight - 451 g"               , ymd("2024-04-14"),
  "left biscuit/gum behind"      , ymd("2024-02-10")
) %>%
  group_by(Date) %>%
  summarise(health_note = str_c(health_note, collapse = "; "), .groups = "drop")
```

## Metadata by Subject and Date

Now I am going to populate metadata tables for each subject by day.

### Culi

```{r, message = FALSE}
culi.samples   <- samples %>% filter(Subject == "Culi")

culi.metadata <- study_days %>%
  left_join(culi.schedule) %>%
  left_join(culi.meds) %>%
  left_join(culi.housing) %>%
  left_join(culi.bristol) %>%
  left_join(access.schedule) %>%
  left_join(culi.repro.notes) %>%
  left_join(culi.health) %>%
  mutate(CollectionDate = Date, .keep = "unused") %>%
  right_join(culi.samples, by = join_by(CollectionDate))
```

### Warble

```{r, message = FALSE, warning = FALSE}
warble.samples <- samples %>% filter(Subject == "Warble")

warble.metadata <- study_days %>%
  left_join(warble.schedule) %>%
  left_join(warble.meds) %>%
  left_join(warble.housing) %>%
  left_join(access.schedule) %>%
  left_join(estrus.schedule) %>%
  left_join(preg.schedule) %>%
  left_join(warble.repro.notes) %>%
  mutate(CollectionDate = Date, .keep = "unused") %>%
  right_join(warble.samples, by = join_by(CollectionDate))
```

## Merge and Export Complete Metadata Table


```{r, message = FALSE}
metadata <- samples %>%
  filter(Subject != "Warble" & Subject != "Culi") %>%
  full_join(warble.metadata) %>%
  full_join(culi.metadata) %>%
  mutate(subj_letter = str_sub(Subject, 1L, 1L)) %>%
  mutate(subj_day    = str_glue("{subj_letter}{study_day}")) %>%
  select(-subj_letter) %>%
  relocate(
    SequenceID,
    ExtractID,
    SampleID,
    study_day,
    subj_day,
    CollectionDate:health_note
  ) %>%
  arrange(CollectionDate, Subject)
```

```{r}
write.table(metadata,
            path$metadata$summary,
            row.names = F,
            sep = "\t")

metadata %>%
  kbl() %>%
  kable_styling(full_width = T,
                fixed_thead = T,
                c("hover", "responsive"),
                position = "center") %>%
  column_spec(1, bold = T) %>%
  collapse_rows(columns = 1, valign = "top") %>%
  scroll_box(width = "800px", height = "800px")

```


### Key for Metadata Variables

I also create a separate R script with a list of every metadata variable and its description. The code below is an easy way to generate a list of all column headers from this very large table that you can copy and paste for writing the list.  

```{r}
print(names(metadata))
```


Now I will bring in the key that I previously created and print.  

```{r}
source(path$metadata$key)

tagList(
  tags$h2("Metadata Variables:"),
imap(metadata_key, ~ {
  var_name  <- .y
  var_descr <- .x
  tags$p(
    tags$strong(var_name),
    tags$br(),
      var_descr
  )
}))
```



# Next Steps

>Now you should proceed to the MicroEco Workflow to Begin Merging Metadata with Microbiome Results for Analysis.





