---
title: "Preprocessing Microbiome Data with Phyloseq"
author: "Alicia M. Rich, Ph.D."
date: "`r Sys.Date()`"
output:
  html_document:
    theme:
      bslib: true
    toc: true
    toc_depth: 3
    toc_float: true
    df_print: kable
    css: journal.css
    code_download: true
  
---

```{r setup, include=FALSE}
library(conflicted)
library(tidyverse)
library(seqinr)
library(ape)
library(treedataverse)
library(phyloseq)
library(Biostrings)
library(paletteer)
library(skimr)

source("setup/conflicted.R")


knitr::opts_chunk$set(
  message       = FALSE,
  warning       = FALSE,
  echo          = TRUE,
  include       = TRUE,
  eval          = TRUE,
  comment       = ""  ,
  df_print      = "kable",
  skimr_digits  = 2        )

knitr::opts_knit$set(
  message       = FALSE,
  warning       = FALSE,
  echo          = TRUE,
  include       = TRUE,
  eval          = TRUE,
  comment       = ""  ,
  df_print      = "kable",
  skimr_digits  = 2        
  )
```


# Preprocessing Overview

In this section we will take raw microbiome count data and turn it into a clean, analysis-ready `phyloseq` object.  You will:

1.	**Load your core inputs**

  - A rooted phylogenetic tree (`refs_tree.treefile`)
    - produced in the [`microbiome_references` tutorial](https://rich-molecular-health-lab.github.io/tutorials/microbiome_references.html).
  - Representative 16S sequences (`refs_aligned_mafft.fasta`)
    - produced in the [`microbiome_references` tutorial](https://rich-molecular-health-lab.github.io/tutorials/microbiome_references.html).
  - A taxonomic assignment table (`tax_table.RData`)
    - produced in the [`microbiome_alignments` tutorial](https://rich-molecular-health-lab.github.io/tutorials/microbiome_alignments.html).
  - Sample metadata (`sample_table.RData`)
    - produced in the [`metadata_setup` tutorial](https://rich-molecular-health-lab.github.io/tutorials/metadata_setup.html) and then prepared for microbiomes in [`microbiome_alignments`]((https://rich-molecular-health-lab.github.io/tutorials/microbiome_alignments.html).
  - An OTU count table (`otu_table.RData`)
    - produced in the [`microbiome_alignments` tutorial](https://rich-molecular-health-lab.github.io/tutorials/microbiome_alignments.html).
    
2.  **Subset Samples**

  - Exclude any samples not produced by male *Culi* so that we can normalize within this subset that we are currently looking at for our case study.
	
3.	**Merge replicate samples**

  - Pool all reads for the same subject (here just Culi) on the same day into a single sample.  OTU counts are summed, and numeric metadata are averaged.

4.	**Clean up taxonomy & prune**

  - Remove very low-abundance OTUs (e.g. total count ≤ 1)
  - Collapse your count table and tree to the Genus rank
  - Perform a second pruning step so that only genera with meaningful counts remain

5.	**Save analysis-ready objects**

  - A cleaned, genus-level phyloseq object (`phyloseq_genus.RData`)
  - Its melted version (`phyloseq_melt_genus.RData`) for ggplot workflows
  - The pruned, genus-level tree (`tree_genus.RData`)
  - Analogous objects for Family, Order, Class, and Phylum

## By the end of this stage, you will have one phyloseq object per taxonomic level, each containing:

  - A consistent OTU (or Genus/Family…) count table
  - Matched sample metadata and taxonomic table
  - A pruned phylogenetic tree
  
# Packages Used

In this script we mostly use the `phyloseq` package, designed for microbiome analysis.

## `phyloseq` Utility

- Integrates OTU/ASV tables, sample metadata, taxonomy, phylogenetic trees, and (optionally) reference sequences in one S4 object.
- Provides convenient methods for merging, pruning, rarefaction, distance calculations, and plotting.
- Under the hood, relies on Bioconductor classes (DataFrames, XStringSets, phylo).

>Note: there is much overlap between `phyloseq`, `microeco`, and `microbiome`, but each makes some steps more streamlined than others, so I prefer to switch between them for different stages of analysis.

We will also use the `treedataverse` package suite for some visualization to check in on our data quality.

## `treedataverse` Utility

The `treedataverse` suite provides tools for managing, sharing, and visualizing phylogenetic trees in R. It builds on the Dataverse platform to store trees (and associated data) in versioned, citable repositories. With `treedataverse`, you can:

  - Upload or download phylo objects directly from a Dataverse collection
  - Attach rich metadata (e.g. study details, taxon provenance) to each tree
  - Track updates and collaborate on tree data with full provenance
  - Integrate seamlessly with ggtree and ape for downstream plotting and analysis

### `ggtree` Package within the `treedataverse`

`ggtree` is an extension of `ggplot2` tailored for phylogenetic trees. It lets you:

  - Map tree data (branches, node labels, support values) to aesthetic scales
  - Annotate clades with heatmaps, bar charts, or tip points (e.g. taxon abundances)
  - Collapse or expand subtrees interactively
  - Align associated data (e.g. trait matrices) alongside your tree in multi-panel layouts
  - Leverage the full grammar of graphics for publication-quality figures

---

# Helper Functions

I created the functions below because I repeatedly use these code chunks to view summaries of my data after each step of processing.

```{r}

my_skim <- skim_with(numeric  = sfl(p25 = NULL, p50 = NULL, p75 = NULL))

skim_phyloseq <- function(phyloseq) {
options(
  digits            = 2,
  pillar.sigfig     = 2,
  pillar.subtle_num = TRUE
  )
 skim <- psmelt(phyloseq) %>% 
    select(-c(
      subject_certainty, 
      subject_sex,
      subject_studbook_id, 
      subject_dam_id, 
      subject_sire_id, 
      subject_birth_location, 
      subject_age)) %>%
    group_by(diet_name) %>% 
    my_skim() %>%
    select(-complete_rate)
 
 return(skim)
}

view_tree <- function(phyloseq, level = NULL, palette = "rcartocolor::Prism") {
  phylum.cols <-  sample(paletteer::paletteer_d(paste0(palette)), size = length(get_taxa_unique(phyloseq, "Phylum")), replace = TRUE) %>%
    set_names(., get_taxa_unique(phyloseq, "Phylum"))
  
  tree1 <- ggtree(phyloseq, branch.length = "none", aes(color = Phylum))
  
  if (is.null(level)) {
    tree2 <- tree1
  } else if (level == "Class") {
    tree2 <- tree1 +
      geom_label(aes(label = Class, fill = Phylum), hjust = -0.05, size = 2, colour = "#000000FF")
  } else if (level == "Phylum") {
    tree2 <- tree1 +
      geom_label(aes(label = Phylum, fill = Phylum), hjust = -0.05, size = 3, colour = "#000000FF")
  } else if (level == "Order") {
    tree2 <- tree1 +
      geom_label(aes(label = Order, fill = Phylum), hjust = -0.05, size = 2, colour = "#000000FF")
  }
  
  tree.out <- tree2 +
    scale_color_manual(values = phylum.cols, na.value = "#00000080") +
    scale_fill_manual( values = alpha(c(phylum.cols), 0.3)) +
    theme_tree(legend.position = "none")
  
  return(tree.out)
}
```

---

# Load Data

>Note that I use `.RData` files to load some of the data below. You could load these same objects in the form of more widely applicable files, if you wish (*see below for explaination*).

<details>
<summary><emph>What is an .RData file?</emph></summary>
<p>An `.RData` (or sometimes `.rda`) file is R’s native, binary workspace format. When you call `save()` on one or more R objects—data frames, matrices, lists, S4/S3 objects, etc.—R serializes them into a compact, platform-independent file. You can later restore those exact objects (with their names and attributes intact) by calling load("`your_file.RData`"). Using `.RData` is an efficient way to cache intermediate steps, share processed datasets with collaborators, or speed up analyses by skipping time-consuming recomputation.</p>
</details>

```{r}
tree <- read.tree("microeco/loris/refs_tree.treefile")
seqs <- readDNAStringSet("microeco/loris/refs_aligned_mafft.fasta")
load("microeco/loris/tax_table.RData")
load("microeco/loris/sample_table.RData")
load("microeco/loris/otu_table.RData")
```

## Sample Metadata as Factors

Phyloseq will only capture numeric/factor values (not strings) for the metadata, so we need to reformat our character columns.

```{r}
sample_data <- sample.table %>%
  arrange(subject, collection_day) %>%
  mutate(subject_certainty       = fct_recode(subject_certainty  , "0" = "no" , "1" = "yes"),
         environment_holding     = fct_recode(environment_holding, "0" = "old", "1" = "new"),
         environment_pair_access = fct_recode(environment_pair_access, "0" = "n", "1" = "y"),
         repro_warb_status       = fct_recode(repro_warb_status, "0" = "anestrus", "1" = "estrus")) %>%
  mutate(across(where(is.character), ~factor(.))) %>%
  mutate(diet_name = fct_inorder(diet_name))
```
  
## Phyloseq Object

```{r}
phylo <- phyloseq(
  tax_table(as.matrix(tax.table)),
  otu_table(as.matrix(otu.table), taxa_are_rows = TRUE),
  sample_data(sample_data),
  phy_tree(tree),
  refseq(seqs)
)
```

## Preview Starting Data

```{r}
skim_phyloseq(phylo)
```


```{r}
view_tree(phylo) + ggtitle("Species-Level OTUs before Filtering")
```

That looks like some very dense, messy taxonomic data!

---

# Subset and Dereplicate Samples

First, we will take a subset of our samples, **keeping only those for Culi**, as I am currently only looking for patterns over time in his gut, not comparing Culi and Warble (yet).

```{r}
phylo_culi <- phylo %>% subset_samples(subject == "culi")
```


Next we will create a version that merges the replicated samples so that we have one observation per day. This function's default in phyloseq will aggregate the OTU counts across multiple samples as a sum of reads while taking mean values for everything in the metadata.

>This pooling step creates **one composite sample per subject per day**, avoiding pseudo‐replication in downstream diversity metrics.

```{r}
phylo_merged <- phylo_culi %>% merge_samples("subject_day")

skim_phyloseq(phylo_merged)
```


# Clean up Taxonomy

## Prune Taxa

Drop the OTUs with the lowest abundances to help stabilize diversity estimates and speed up tree operations. We will constrain based on:

  1.  Total sum of OTU counts across all samples
  2.  Proportion of samples where OTU is present
  
```{r}
# custom function for filtering thresholds

keep_taxa <- function(phyloseq, tax_thresh, samp_thresh) {
  taxa_names(phyloseq)[
  taxa_sums(phyloseq) > tax_thresh &
  apply(otu_table(phyloseq), 1, function(x) sum(x > 0)) > nsamples(phyloseq)*samp_thresh
]
}

report_threshold <- function(phyloseq, keep) {
  cat(paste0(
    "starting with ", ntaxa(phyloseq), " OTUs, discarding ", (ntaxa(phyloseq) - length(keep)), " OTUs, keeping ", length(keep), " OTUs"
  ))
}
```

  
**First let's see how many OTUs we lose at different thresholds**

```{r}
keep_otus <- keep_taxa(phylo_merged, 10, 0.1)
report_threshold(phylo_merged, keep_otus)
```


```{r}
keep_otus <- keep_taxa(phylo_merged, 5, 0.05)
report_threshold(phylo_merged, keep_otus)
```

For now, we will remove 1,574 species where the total abundance is less than or equal to 5 or the species only appears in less than 5% of all samples. (*You can change the value in the `keep_taxa` function above to use a higher/lower threshold.*)

```{r}
phylo_pruned <- prune_taxa(keep_otus, phylo_merged)
skim_phyloseq(phylo_pruned)
```

```{r}
view_tree(phylo_pruned) + ggtitle("Species-Level OTUs after Filtering")
```


## Collapse Taxa to Genus Level

Collapsing at genus level reduces complexity and allows genus‐scale ecological patterns to emerge with less interference from random error at the species-assignment level.

```{r}
phylo_genus <- phylo_pruned %>% tax_glom("Genus")
skim_phyloseq(phylo_genus)
```


```{r}
view_tree(phylo_genus) + ggtitle("Genus-Level OTUs")
```

## Another Pruning Check

Let's look at our OTUs now that we merged to the Genus level and consider filtering again at this level.


```{r}
keep_genera <- keep_taxa(phylo_genus, 10, 0.1)
report_threshold(phylo_genus, keep_genera)
```

Let's do one more pruning step to only keep the genera with total abundances greater than 10 and presence in at least 10% of all samples (233 genera total).

```{r}
culi_genus <- prune_taxa(keep_genera, phylo_genus)
skim_phyloseq(culi_genus)
```

```{r results='asis'}
view_tree(culi_genus) + ggtitle("Genus-Level OTUs")
```


## Extract and Save Clean Phyloseq Objects

Exporting intermediate objects saves time if you need to re‐run later steps without re‐processing.

```{r}
culi_melt_genus     <- psmelt(culi_genus)
culi_tree           <- phy_tree(culi_genus)
save(culi_genus     , file = "microeco/loris/culi/culi_genus.RData")
save(culi_tree      , file = "microeco/loris/culi/culi_tree.RData")
save(culi_melt_genus, file = "microeco/loris/culi/culi_melt_genus.RData")
```

# Subset Additional Taxonomic Levels

Working at multiple ranks lets you compare community structure from broad (Phylum) to fine (Genus) scales in the same pipeline.

```{r}
fam_culi <- culi_genus %>% tax_glom("Family")
ord_culi <- culi_genus %>% tax_glom("Order")
cla_culi <- culi_genus %>% tax_glom("Class")
phy_culi <- culi_genus %>% tax_glom("Phylum")

view_tree(fam_culi) + ggtitle("Family-Level OTUs") 
view_tree(ord_culi, "Order") + ggtitle("Order-Level OTUs") + 
      coord_cartesian(clip = "off") + 
      theme(plot.margin = margin(5, 65, 5, 5, "pt"))
view_tree(cla_culi, "Class") + ggtitle("Class-Level OTUs") + 
      coord_cartesian(clip = "off") + 
      theme(plot.margin = margin(5, 65, 5, 5, "pt"))
view_tree(phy_culi, "Phylum") + ggtitle("Phylum-Level OTUs") +  
      coord_cartesian(clip = "off") + 
      theme(plot.margin = margin(5, 65, 5, 5, "pt"))
```

## Extract and Save for Other Levels

```{r}
fam_culi_melt     <- psmelt(fam_culi)
ord_culi_melt     <- psmelt(ord_culi)
cla_culi_melt     <- psmelt(cla_culi)
phy_culi_melt     <- psmelt(phy_culi)

fam_culi_tree           <- phy_tree(fam_culi)
ord_culi_tree           <- phy_tree(ord_culi)
cla_culi_tree           <- phy_tree(cla_culi)
phy_culi_tree           <- phy_tree(phy_culi)

save(fam_culi     , file = "microeco/loris/culi/fam_culi.RData")
save(ord_culi     , file = "microeco/loris/culi/ord_culi.RData")
save(cla_culi     , file = "microeco/loris/culi/cla_culi.RData")
save(phy_culi     , file = "microeco/loris/culi/phy_culi.RData")

save(fam_culi_tree      , file = "microeco/loris/culi/fam_culi_tree.RData")
save(ord_culi_tree      , file = "microeco/loris/culi/ord_culi_tree.RData")
save(cla_culi_tree      , file = "microeco/loris/culi/cla_culi_tree.RData")
save(phy_culi_tree      , file = "microeco/loris/culi/phy_culi_tree.RData")

save(fam_culi_melt, file = "microeco/loris/culi/fam_culi_melt.RData")
save(ord_culi_melt, file = "microeco/loris/culi/ord_culi_melt.RData")
save(cla_culi_melt, file = "microeco/loris/culi/cla_culi_melt.RData")
save(phy_culi_melt, file = "microeco/loris/culi/phy_culi_melt.RData")
```

---

# Next Steps

In the [`microbiome_metrics` tutorial](https://rich-molecular-health-lab.github.io/tutorials/microbiome_metrics.html) you will learn how to choose and apply rarefaction or relative-abundance normalization, and then calculate fundamental diversity statistics (α-diversity indices, β-diversity distances, UniFrac, etc.).  Refer to that section to decide when and how to rarefy your merged tables, or to work directly on normalized counts for non-phylogenetic metrics.



