---
title: "Prep Sample Sheets for MinION"
author: "Alicia Rich"
output:
  html_document:
    toc: true
    toc_location: "before"
    toc_depth: 4
    number_sections: false
    toc_float: true
    code_folding: "hide"
    fig_caption: true
params:
  taxon: "loris"
editor_options: 
  chunk_output_type: inline
                     
---

```{r, setup, include = F}
global             <- config::get(config = "default")
micro              <- config::get(config = "microbiome")
read_processing    <- config::get(config = "read_processing")
notebooks          <- config::get(config = "notebooks")
loris              <- config::get(config = "loris")
marm               <- config::get(config = "marmoset")
path               <- config::get(config = params$taxon)
barcode_alignments <- config::get(config = "barcode_alignments")
sample_sheets      <- config::get(config = "sample_sheets")

day1        <- loris$day1

source(paste0(global$functions))
source(paste0(global$packages))
source(paste0(global$conflicts))
source(paste0(global$inputs))
source(paste0(read_processing$functions))

subject_list <- keep_at(subjects, paste0(params$taxon)) %>% list_flatten(name_spec = "{inner}")
```

# Intro

>This is a very quick and basic script to organize sample lists and sheets for sequencing runs for our Loris Microbiome Project. You should update the sample inventory each time you want to begin processing another set of reads from the MinION. 

**Workflow of Scripts After This One:**  
1. Read Processing  
2. Metadata Loris  
3. MicroEco Workflow  

## Tables to Load into R Studio

>This script will scan your working directory for all files ending in _barcode_alignment.tsv, so you should transfer the barcode alignment output from each sequencing run and add a sequencing run identifier to the beginning of the file name and make sure the rest matches. Then the script will automatically compile all versions in your working directory each time you run the script. 
   
>You also need to have an up-to-date copy of the 16S Barcoding Spreadsheet with complete sample info. Before running this script, I always check the 16S Libraries sheet from the shared google drive to ensure all sample info is up to date. If you download that as a CSV file as is and transfer it to your working directory, it should work in this script.  

### List of files:
<ol>
  <li>_barcode_alignment.tsv</li>
  <li>HDZ_Loris_16S_Barcoding_*.csv</li>
</ol>

# Barcode Alignment Tables

>Make sure the maximum number for each taxon list below includes the highest number of sequencing run completed so far (as of now we are up to 18 sequencing runs for hdz/lorises). That will ensure you load the complete list of files for whichever sampleset you are working with.

```{r}
seqruns <- list(
  loris     = as.list(paste0("hdz", 1:18)),
  marmoset  = as.list(sprintf("cm%03d", 1:10))
)
```


```{r}
barcodes <- imap(seqruns[[paste0(params$taxon)]], ~ {
  map(.x, ~ read.table(barcode_alignments[[.x]], header = T) %>% mutate(seqrun = .x)) 
}) %>%
    bind_rows() %>%
                        as_tibble() %>%
                        filter(barcode     != "unclassified") %>%
                        mutate(SeqDateTime  = as_datetime(started)) %>%
                        mutate(SeqDate      = floor_date(SeqDateTime, unit = "day")) %>%
                        mutate(SeqRunID     = str_replace_all(sample_id, "pool1", "PL001")) %>%
                      mutate(LibraryCode    = str_squish(str_trim(seqrun      , "both")),
                             LibraryBarcode = str_squish(str_trim(barcode     , "both")),
                             FlowCellSerial = str_squish(str_trim(flow_cell_id, "both"))
                             ) %>%
                      mutate(LibraryBarcode = str_remove_all(LibraryBarcode, "barcode0")) %>%
                      mutate(LibraryBarcode = as.numeric(str_remove_all(LibraryBarcode, "barcode"))) %>%
                        select(LibraryCode,
                               LibraryBarcode,
                               reads_unclassified = target_unclassified,
                               FlowCellSerial,
                               protocol_group_id,
                               SeqRunID,
                               SeqDate,
                               SeqDateTime)

write.table(barcodes, barcode_alignments$compilations[[paste0(params$taxon)]],
            row.names = F,
            sep = "\t")
```

# Sample Collection Records

```{r}
samples     <- read.csv(path$samples_csv, header = T) %>% 
                      filter(str_starts(SampleID, "\\w+")) %>% 
                      select(-SampleBox)  %>%
                      mutate(SampleID = str_squish(str_trim(SampleID, "both"))) %>% distinct() %>%
                      mutate(CollectionDate     = mdy(SampleDate),
                             Subject            = str_squish(str_trim(SampleSubject)),
                             .keep = "unused") %>% distinct() %>%
                      mutate(Subj_Certainty = if_else(Subject %in% subject_list, "yes", "no")) %>%
                      mutate(Subject        = str_remove_all(Subject, "\\?"))

extracts <- read.csv(path$extracts_csv, header = T) %>% 
  filter(str_starts(SampleID, "\\w+")) %>%
  mutate(SampleID = str_squish(str_trim(SampleID, "both")),
         ExtractID= str_squish(str_trim(ExtractID, "both")),
         ExtractDate       = mdy(ExtractDate)) %>%
  mutate(ExtractConc       = str_remove_all(ExtractConcentration, ">"), .keep = "unused") %>%
  mutate(ExtractConc = if_else(str_detect(ExtractConc, "Higher"), "100", ExtractConc),
         ExtractConc = if_else(str_detect(ExtractConc, "HIGHER"), "100", ExtractConc),
         ExtractConc = if_else(ExtractConc == "LOW", "0", ExtractConc),
         ExtractConc = if_else(ExtractConc == "", NA, ExtractConc)) %>%
  mutate(ExtractConc = round(as.numeric(ExtractConc), 1))  %>% filter(ExtractType == "DNA") %>%
  select(-ExtractType) %>%
  left_join(samples) %>% distinct()
```


```{r}
compilation <- read.csv(path$compilation_csv, header = T)      %>% 
  filter(str_starts(SampleID, "\\w+") & str_starts(SequenceID, "\\w+") & Pooled.Library.Code != "#N/A") %>% 
  mutate(SampleID   = str_squish(str_trim(SampleID  , "both")),
         ExtractID  = str_squish(str_trim(ExtractID , "both")),
         SequenceID = str_squish(str_trim(SequenceID, "both"))) %>%
  mutate(SampleSet      = paste(params$taxon),
         LibraryTube    = as.integer(LibraryTube),
         LibPrepDate    = ymd(str_remove_all(str_trim(Run.ID, "both"), "MIN_16_|MIN_16-")),
         strands        = 2,
         fragment_type  = 0,
         Length         = 1500,
         InputMassStart = 10,
         SampVolPool    = round(as.numeric(Volume.Added.to.Pool..uL.), 0)) %>% 
         mutate(LibraryBarcode = str_remove_all(LibraryBarcode, "16S")) %>%
         mutate(LibraryBarcode = str_remove_all(LibraryBarcode, "barcode0")) %>%
         mutate(LibraryBarcode = as.numeric(str_remove_all(LibraryBarcode, "barcode"))) %>%
  select(-c(ExtractDate:SampleBox)) %>%
  left_join(extracts, by = join_by(ExtractID)) %>% distinct() %>%
  mutate(LibPrepWorkflow = if_else(str_detect(Kit, "LSK"), "lsk16s", "rapid16s")) %>%
  mutate(TemplateVolPrep = if_else(LibPrepWorkflow == "lsk", 47, 15),
         PoolSamples     = "yes",
         InputMassFinal  = 50,
         SeqRunID        = Pooled.Library.Code) %>%
  mutate(LibraryCode     = case_when(
    SampleSet == "loris" & str_detect(SeqRunID, "PL00") ~ str_replace_all(SeqRunID, "PL00", "hdz"),
    SampleSet == "loris" & str_detect(SeqRunID, "PL01") ~ str_replace_all(SeqRunID, "PL0" , "hdz"),
    SampleSet == "marmoset" ~ str_to_lower(SeqRunID),
    .default = SeqRunID)) %>%
  mutate(TotalPoolVol = sum(SampVolPool), .by = LibraryCode) %>%
  mutate(BeadVol = TotalPoolVol * 0.6) %>%
  select(LibraryTube,
         ExtractID,
         SampleID,
         LibraryBarcode,
         ExtractConc,
         SampleSet,
         LibraryCode,
         SequenceID,
         LibPrepDate,
         LibPrepWorkflow,
         LibPrepKit = Kit,
         FlowCellType   = Flow.Cell.Type,
         SeqDevice = Sequencer,
         strands,
         fragment_type,
         Length,
         InputMassStart,
         TemplateVolPrep,
         PoolSamples,
         SampVolPool,
         TotalPoolVol,
         BeadVol,
         InputMassFinal,
         Conc_QC2 = Final.Library.Concentration) %>%
  arrange(ExtractID, SequenceID) %>% distinct() %>%  
  left_join(barcodes, by = join_by(LibraryCode, LibraryBarcode)) %>%
  arrange(LibraryCode)
```


```{r}
count.extracts    <- extracts %>% select(ExtractID, SampleID) %>% distinct() %>% 
  group_by(SampleID)  %>% 
  mutate(n_dna_extracts = n_distinct(ExtractID)) %>% ungroup() %>% select(-ExtractID)

count.libraries <- compilation %>% select(SequenceID, ExtractID, SampleID) %>% distinct() %>% 
  group_by(ExtractID) %>% mutate(n_16s_extract = n_distinct(SequenceID)) %>% ungroup() %>%
  group_by(SampleID)  %>% mutate(n_16s_sample  = n_distinct(SequenceID)) %>% ungroup() %>% select(-SequenceID)
```

```{r}
samplesheet <- compilation %>%
                      mutate(barcode = if_else(LibraryBarcode < 10, 
                                               str_glue("barcode0", "{LibraryBarcode}"),
                                               str_glue("barcode" , "{LibraryBarcode}"))) %>%
                      select(flow_cell_id  = FlowCellSerial,
                             experiment_id = protocol_group_id,
                             kit           = LibPrepKit,
                             barcode,
                             alias         = SequenceID,
                             seqrun        = LibraryCode)

write.table(samplesheet, 
          sample_sheets$compilations[[paste0(params$taxon)]],
          row.names = F,
          quote     = F,
          sep       = ",")
```


```{r}
samplesheet.nested <- samplesheet %>% mutate(LibraryCode = seqrun) %>% nest(.by = LibraryCode) %>%
  deframe()

imap(samplesheet.nested, ~ {
  write.table(.x, (sample_sheets[[.y]]),
          row.names = F,
          quote     = F,
          sep       = ",")
})
```



# Next Step

>Now you should proceed to the Read Processing workflow to begin basecalling the sequencing run.
