---
title: "Microbiome Results for Culi Only - IN PROGRESS"
author: "Alicia Rich"
output:
  html_document:
    theme:
      bootswatch: litera
    toc: true
    toc_float:
      collapsed: true
      smooth_scroll: true
    code_folding: "show"
    fig_caption: true
    df_print: paged
params:
  sampleset: "loris"
  dataset: "culi"
                     
---

```{r global, message = FALSE}
global             <- config::get(config = "default")

here::i_am("Results_CuliMicrobiome.Rmd")
source(here::here(global$setup))

for (file in micro$micro_scripts) {
  source(here(file))
}

source(here(path$metadata$key))
source(here(path$metadata$factors))

dataset_dir <- path$microeco$dataset[paste0(params$dataset)] %>% list_flatten(name_spec = "{inner}")
abund_dir <- path$microeco$abund[paste0(params$dataset)] %>% list_flatten(name_spec = "{inner}")

theme_set(theme_classic())
thematic_rmd()
thematic_on(accent = "#8785B2FF", fg = "black")
```

# Read in and Format Datasets

I have written several custom functions to expedite some of the functions that I reuse multiple times for data sets across different sample and taxon subsets. The loop below will use one of those functions to read the microeco datasets that I built using the [MicroEcoDataPrep](MicroEcoDataPrep.html) script.  

This code makes use of the dataset paths referenced in my config.yaml file. To see the full path (relative to the repository base directory), refer to the relative syntax in this script and the config file below.  

```{r, echo = FALSE}
page_fluid(
    accordion(
      open = FALSE,
      accordion_panel(
        "Show/Hide Config File (config.yml)",
        tagList(tags$pre(includeText("config.yml")))
    ),
      accordion_panel(
        "Show/Hide Variables and Lists Used in Functions",
        tagList(
          withTags(
            ul(
              li(em("dataset_types")        , pre(toString(dataset_types))),
              li(em("tax_levels")           , pre(toString(tax_levels))),
              li(em("datasets")             , pre(toString(datasets))),
              li(em("func_levels")          , pre(toString(func_levels))),
              li(em("diet_factors")         , pre(toString(diet_factors))),
              li(em("warble_cycle_factors") , pre(toString(warb_cycle_factors))),
              li(em("holding_factors")      , pre(toString(holding_factors))),
              li(em("probiotic_factors")    , pre(toString(probiotic_factors))),
              li(em("steroid_factors")      , pre(toString(steroid_factors))),
              li(em("fiber_factors")        , pre(toString(fiber_factors))),
              li(em("antibiotic_factors")   , pre(toString(antibiotic_factors))),
              li(em("antidiarrheal_factors"), pre(toString(antidiarrheal_factors))),
              li(em("sample_table_numeric") , pre(toString(sample_table_numeric))),
              li(em("env_cols")             , pre(toString(env_cols)))
              )
            )
          )
        )
    )
  )
```


```{r}
sample_table <- read.table(path$metadata$sample_table$merged, 
                           header = T, 
                           sep = "\t") %>%
  mutate(ID = str_glue("{subject}", "_", "{study_day}")) %>%
  relocate(ID) %>%
  mutate(subject      = fct(subject),
         pair_access  = fct(pair_access  , levels = c("n", "y")),
         diet_name    = fct(diet_name    , levels = diet_factors),
         warb_status  = fct(warb_status  , levels = warb_cycle_factors),
         holding      = fct(holding      , levels = holding_factors))

culi.bristols <- sample_table %>%
  filter(subject == "culi", bristol_mean > 0) %>%
  summarize(med = median(bristol_mean, na.rm = TRUE))

dataset_tbl <- sample_table %>%
  mutate(bristol_mean = if_else(is.na(bristol_mean) | bristol_mean == 0, culi.bristols$med, bristol_mean)) %>%
  filter(subject == "culi") %>%
  select(
 "ID"                         ,
 "study_day"                  ,
 "diet_name"                  ,
 "total_mg"                   ,
 "total_kcal"                 ,
 "total_mg_dry"               ,
 "probiotic"                  ,
 "fiber"                      ,
 "steroid"                    ,
 "antibiotic"                 ,
 "antidiarrheal"              ,
 "bristol_mean"               ,
 "holding"                    ,
 "Biscuit"                    ,
 "Gum_Arabic"                 ,
 "Invertebrates"              ,
 "Protein_Rotation"           ,
 "Seasonal_Veggie_Rotation"   ,
 "Oat_Gel"                    ,
 "CollectionDate"             ,
 "protein_fed"                ,
 "fat_fed"                    ,
 "CHO_fed"                    ,
 "mineral_fed"                ,
 "Omega_3"                    ,
 "Omega_6"                    ,
 "Methionine"                 ,
 "Taurine"                    ,
 "Calcium_Ca"                 ,
 "Magnesium_Mg"               ,
 "Phosphorus_P"               ,
 "Potassium_K"                ,
 "Copper_Cu"                  ,
 "Iodine_I"                   ,
 "Iron_Fe"                    ,
 "Manganese_Mn"               ,
 "Zinc_Zn"                    ,
 "Acid_Detergent_Fiber"       ,
 "Neutral_Detergent_Fiber"    ,
 "Total_Dietary_Fiber"        ,
 "Water_Soluble_Carbohydrates",
 "Crude_Fiber"                ,
 "Starch"                     ,
 "Beta_Carotene"              ,
 "Lycopene"                   ,
 "Choline"                    ,
 "Folic_Acid_Vitamin_B9"      ,
 "Vitamin_B1_Thiamin"         ,
 "Vitamin_B2_Riboflavin"      ,
 "Vitamin_B3_Niacin"          ,
 "Vitamin_B5_Pantothenic_Acid",
 "Vitamin_B6_Pyridoxine"      ,
 "Vitamin_B7_Biotin"          ,
 "Vitamin_B12_Cobalamin"      ,
 "Vitamin_C_Ascorbic_Acid"    ,
 "Vitamin_A"                  ,
 "Vitamin_D3_Cholecalciferol" ,
 "Vitamin_E"                  ,
 "Vitamin_K"                  
  ) %>%
  column_to_rownames("ID")
tax_table   <- read.table(paste0(dataset_dir$species, "/tax_table.tsv"), sep = "\t", header = T, row.names = 1)
otu_table   <- read.table(paste0(dataset_dir$species, "/feature_table.tsv"), sep = "\t", header = T, row.names = 1)
rep_fasta   <- read.fasta(paste0(dataset_dir$species, "/rep_fasta.fasta"))
phylo_tree   <- read.tree(paste0(dataset_dir$species, "/phylo_tree.tre"))
```

```{r}
culi <- microtable$new(
          otu_table    = otu_table,
          sample_table = dataset_tbl,
          tax_table    = tax_table,
          phylo_tree   = phylo_tree,
          rep_fasta    = rep_fasta,
          auto_tidy    = TRUE
)
```

## Calculate Basic Statistics for Interpretation


```{r}
culi$cal_abund()
culi$filter_taxa(
    rel_abund      = methods_16s$loris$min_abund, 
    freq           = methods_16s$loris$min_freq, 
    include_lowest = methods_16s$loris$include_lowest
  )
culi$cal_alphadiv(PD = methods_16s$loris$alpha_pd)

culi$cal_betadiv(
    unifrac = methods_16s$loris$unifrac, 
    method  = methods_16s$loris$betadiv
  )

```

# Time Series Analysis

## Data Formatting for BEEM Package

```{r}
culi.beem <- clone(culi)

culi.beem$filter_taxa(
    rel_abund      = 0.005, 
    freq           = 0.1, 
    include_lowest = FALSE
  )

culi.abund <- trans_abund$new(
  dataset = culi.beem,
  taxrank = "Genus",
  show    = 0.5
)
  
```

>From the package:
>> The **OTU Table** should be a tab-delimited text file whose first row has the sample IDs and the first column has the OTU IDs (or taxonomic annotations). Each row should then contain the relative abundance of one OTU across all samples and each column should contain the relative abundances of all OTUs in that sample.  
>>The metadata file should be a tab-delimited text file with the following columns:
```
sampleID    isIncluded    subjectID    measurementID
```
 - `sampleID`: sample IDs matching the first row of the OTU table
 - `isIncluded`: whether the sample should be included in the analysis (1-include, 0-exclude)
 - `subjectID`: indicator for which biological replicate the sample belongs to
 - `measurementID`: time in standardized units from the start of the experiment

```{r}
beem.otu <- culi.abund$data_abund %>%
  mutate(Abundance = Abundance * 100) %>%
  select(Genus         = Taxonomy,
         Abundance,
         sampleID      = Sample) %>%
  pivot_wider(id_cols     = "Genus",
              names_from  = sampleID,
              values_from = Abundance) %>%
  column_to_rownames("Genus")

beem.meta <- culi.abund$data_abund %>%
  select(sampleID      = Sample,
         measurementID = study_day
         ) %>%
  distinct() %>%
  mutate(isIncluded = 1,
         subjectID  = "culi")
```


## Run BEEM

```{r}
beem.res <- EM(
  dat  = beem.otu,
  meta = beem.meta,
  ncpu = 3
)
```




